// @progress/kendo-angular-upload/dist/es/rendering/upload-status-total.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, Input } from '@angular/core';
import { FileMap } from '../types/file-map';
import { FileState } from '../types';
import { LocalizationService } from '@progress/kendo-angular-l10n';
/**
 * @hidden
 */
var UploadStatusTotalComponent = /** @class */ (function () {
    function UploadStatusTotalComponent(localization) {
        this.localization = localization;
    }
    UploadStatusTotalComponent.prototype.ngDoCheck = function () {
        this.isPaused = this.fileList.hasFileWithState([FileState.Paused]);
        this.isFailed = this.fileList.hasFileWithState([FileState.Failed]);
        this.isUploading = this.fileList.hasFileWithState([FileState.Uploading]);
        if (this.isPaused && !this.isUploading) {
            this.statusText = this.localization.get('headerStatusPaused');
        }
        else {
            this.statusText = this.isUploading ? this.localization.get('headerStatusUploading')
                : this.localization.get('headerStatusUploaded');
        }
        // this.isPaused = this.isFailed = this.isUploading = false;
        // let currentlyUploading: boolean = this.fileList.hasFileWithState([FileState.Uploading]);
        // if (this.fileList.hasFileWithState([FileState.Paused])) {
        //     this.isPaused = true;
        //     this.statusText = this.localization.get('headerStatusPaused');
        // }
        // if (this.isUploading !== currentlyUploading) {
        //     this.isUploading = currentlyUploading;
        //     this.isFailed = this.fileList.hasFileWithState([FileState.Failed]);
        //     this.statusText = this.isUploading ? this.localization.get('headerStatusUploading')
        //                                        : this.localization.get('headerStatusUploaded');
        // }
    };
    UploadStatusTotalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'kendo-upload-status-total',
                    template: "\n        <strong class=\"k-upload-status k-upload-status-total\">\n          <span class=\"k-icon\"\n            [ngClass]=\"{\n                'k-i-checkmark': !this.isUploading && !this.isFailed,\n                'k-i-exception': !this.isUploading && this.isFailed,\n                'k-i-loading': this.isUploading,\n                'k-i-pause-sm': this.isPaused\n            }\">\n          </span>\n          {{statusText}}\n        </strong>\n    "
                },] },
    ];
    /** @nocollapse */
    UploadStatusTotalComponent.ctorParameters = function () { return [
        { type: LocalizationService }
    ]; };
    UploadStatusTotalComponent.propDecorators = {
        fileList: [{ type: Input }]
    };
    return UploadStatusTotalComponent;
}());
export { UploadStatusTotalComponent };
