// @progress/kendo-angular-upload/dist/es2015/rendering/upload-status-total.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, Input } from '@angular/core';
import { FileMap } from '../types/file-map';
import { FileState } from '../types';
import { LocalizationService } from '@progress/kendo-angular-l10n';
/**
 * @hidden
 */
export class UploadStatusTotalComponent {
    constructor(localization) {
        this.localization = localization;
    }
    ngDoCheck() {
        this.isPaused = this.fileList.hasFileWithState([FileState.Paused]);
        this.isFailed = this.fileList.hasFileWithState([FileState.Failed]);
        this.isUploading = this.fileList.hasFileWithState([FileState.Uploading]);
        if (this.isPaused && !this.isUploading) {
            this.statusText = this.localization.get('headerStatusPaused');
        }
        else {
            this.statusText = this.isUploading ? this.localization.get('headerStatusUploading')
                : this.localization.get('headerStatusUploaded');
        }
        // this.isPaused = this.isFailed = this.isUploading = false;
        // let currentlyUploading: boolean = this.fileList.hasFileWithState([FileState.Uploading]);
        // if (this.fileList.hasFileWithState([FileState.Paused])) {
        //     this.isPaused = true;
        //     this.statusText = this.localization.get('headerStatusPaused');
        // }
        // if (this.isUploading !== currentlyUploading) {
        //     this.isUploading = currentlyUploading;
        //     this.isFailed = this.fileList.hasFileWithState([FileState.Failed]);
        //     this.statusText = this.isUploading ? this.localization.get('headerStatusUploading')
        //                                        : this.localization.get('headerStatusUploaded');
        // }
    }
}
UploadStatusTotalComponent.decorators = [
    { type: Component, args: [{
                selector: 'kendo-upload-status-total',
                template: `
        <strong class="k-upload-status k-upload-status-total">
          <span class="k-icon"
            [ngClass]="{
                'k-i-checkmark': !this.isUploading && !this.isFailed,
                'k-i-exception': !this.isUploading && this.isFailed,
                'k-i-loading': this.isUploading,
                'k-i-pause-sm': this.isPaused
            }">
          </span>
          {{statusText}}
        </strong>
    `
            },] },
];
/** @nocollapse */
UploadStatusTotalComponent.ctorParameters = () => [
    { type: LocalizationService }
];
UploadStatusTotalComponent.propDecorators = {
    fileList: [{ type: Input }]
};
