// ngx-videogular/esm5/lib/core/vg-player/vg-player.js
import { __decorate } from "tslib";
import { Output, Component, EventEmitter, ElementRef, HostBinding, QueryList, AfterContentInit, ContentChildren, ViewEncapsulation, OnDestroy } from '@angular/core';
import { VgAPI } from '../services/vg-api';
import { VgFullscreenAPI } from '../services/vg-fullscreen-api';
import { VgUtils } from '../services/vg-utils';
import { VgMedia } from '../vg-media/vg-media';
import { VgControlsHidden } from '../services/vg-controls-hidden';
var VgPlayer = /** @class */ (function () {
    function VgPlayer(ref, api, fsAPI, controlsHidden) {
        this.api = api;
        this.fsAPI = fsAPI;
        this.controlsHidden = controlsHidden;
        this.isFullscreen = false;
        this.isNativeFullscreen = false;
        this.areControlsHidden = false;
        this.onPlayerReady = new EventEmitter();
        this.onMediaReady = new EventEmitter();
        this.subscriptions = [];
        this.elem = ref.nativeElement;
        this.api.registerElement(this.elem);
    }
    VgPlayer.prototype.ngAfterContentInit = function () {
        var _this = this;
        this.medias.toArray().forEach(function (media) {
            _this.api.registerMedia(media);
        });
        this.fsAPI.init(this.elem, this.medias);
        this.subscriptions.push(this.fsAPI.onChangeFullscreen.subscribe(this.onChangeFullscreen.bind(this)));
        this.subscriptions.push(this.controlsHidden.isHidden.subscribe(this.onHideControls.bind(this)));
        this.api.onPlayerReady(this.fsAPI);
        this.onPlayerReady.emit(this.api);
    };
    VgPlayer.prototype.onChangeFullscreen = function (fsState) {
        if (!this.fsAPI.nativeFullscreen) {
            this.isFullscreen = fsState;
            this.zIndex = fsState ? VgUtils.getZIndex().toString() : 'auto';
        }
        else {
            this.isNativeFullscreen = fsState;
        }
    };
    VgPlayer.prototype.onHideControls = function (hidden) {
        this.areControlsHidden = hidden;
    };
    VgPlayer.prototype.ngOnDestroy = function () {
        this.subscriptions.forEach(function (s) { return s.unsubscribe(); });
    };
    VgPlayer.ctorParameters = function () { return [
        { type: ElementRef },
        { type: VgAPI },
        { type: VgFullscreenAPI },
        { type: VgControlsHidden }
    ]; };
    __decorate([
        HostBinding('class.fullscreen')
    ], VgPlayer.prototype, "isFullscreen", void 0);
    __decorate([
        HostBinding('class.native-fullscreen')
    ], VgPlayer.prototype, "isNativeFullscreen", void 0);
    __decorate([
        HostBinding('class.controls-hidden')
    ], VgPlayer.prototype, "areControlsHidden", void 0);
    __decorate([
        HostBinding('style.z-index')
    ], VgPlayer.prototype, "zIndex", void 0);
    __decorate([
        Output()
    ], VgPlayer.prototype, "onPlayerReady", void 0);
    __decorate([
        Output()
    ], VgPlayer.prototype, "onMediaReady", void 0);
    __decorate([
        ContentChildren(VgMedia)
    ], VgPlayer.prototype, "medias", void 0);
    VgPlayer = __decorate([
        Component({
            selector: 'vg-player',
            encapsulation: ViewEncapsulation.None,
            template: "<ng-content></ng-content>",
            providers: [VgAPI, VgFullscreenAPI, VgControlsHidden],
            styles: ["\n        vg-player {\n            font-family: 'videogular';\n            position: relative;\n            display: flex;\n            width: 100%;\n            height: 100%;\n            overflow: hidden;\n            background-color: black;\n        }\n        vg-player.fullscreen {\n            position: fixed;\n            left: 0;\n            top: 0;\n        }\n        vg-player.native-fullscreen.controls-hidden {\n            cursor: none;\n        }\n    "]
        })
        // tslint:disable:component-class-suffix
        // tslint:disable:no-output-on-prefix
    ], VgPlayer);
    return VgPlayer;
}());
export { VgPlayer };
