// ngx-videogular/esm5/lib/core/vg-media/vg-media-element.js
var VgMediaElement = /** @class */ (function () {
    function VgMediaElement() {
    }
    Object.defineProperty(VgMediaElement.prototype, "audioTracks", {
        get: function () { return null; },
        enumerable: true,
        configurable: true
    });
    // @ts-ignore
    VgMediaElement.prototype.addTextTrack = function (kind, label, language) {
        return null;
    };
    // @ts-ignore
    VgMediaElement.prototype.canPlayType = function (type) {
        return null;
    };
    VgMediaElement.prototype.load = function () {
    };
    VgMediaElement.prototype.msClearEffects = function () {
    };
    VgMediaElement.prototype.msGetAsCastingSource = function () {
        return null;
    };
    // @ts-ignore
    VgMediaElement.prototype.msInsertAudioEffect = function (activatableClassId, effectRequired, config) {
    };
    // @ts-ignore
    VgMediaElement.prototype.msSetMediaKeys = function (mediaKeys) {
    };
    // @ts-ignore
    VgMediaElement.prototype.msSetMediaProtectionManager = function (mediaProtectionManager) {
    };
    VgMediaElement.prototype.pause = function () {
    };
    VgMediaElement.prototype.play = function () {
        return null;
    };
    // @ts-ignore
    VgMediaElement.prototype.setMediaKeys = function (mediaKeys) {
        return null;
    };
    // @ts-ignore
    VgMediaElement.prototype.addEventListener = function (type, listener, useCapture) {
    };
    return VgMediaElement;
}());
export { VgMediaElement };
