// ngx-videogular/esm2015/lib/core/vg-player/vg-player.js
import { __decorate } from "tslib";
import { Output, Component, EventEmitter, ElementRef, HostBinding, QueryList, AfterContentInit, ContentChildren, ViewEncapsulation, OnDestroy } from '@angular/core';
import { VgAPI } from '../services/vg-api';
import { VgFullscreenAPI } from '../services/vg-fullscreen-api';
import { VgUtils } from '../services/vg-utils';
import { VgMedia } from '../vg-media/vg-media';
import { VgControlsHidden } from '../services/vg-controls-hidden';
let VgPlayer = 
// tslint:disable:component-class-suffix
// tslint:disable:no-output-on-prefix
class VgPlayer {
    constructor(ref, api, fsAPI, controlsHidden) {
        this.api = api;
        this.fsAPI = fsAPI;
        this.controlsHidden = controlsHidden;
        this.isFullscreen = false;
        this.isNativeFullscreen = false;
        this.areControlsHidden = false;
        this.onPlayerReady = new EventEmitter();
        this.onMediaReady = new EventEmitter();
        this.subscriptions = [];
        this.elem = ref.nativeElement;
        this.api.registerElement(this.elem);
    }
    ngAfterContentInit() {
        this.medias.toArray().forEach((media) => {
            this.api.registerMedia(media);
        });
        this.fsAPI.init(this.elem, this.medias);
        this.subscriptions.push(this.fsAPI.onChangeFullscreen.subscribe(this.onChangeFullscreen.bind(this)));
        this.subscriptions.push(this.controlsHidden.isHidden.subscribe(this.onHideControls.bind(this)));
        this.api.onPlayerReady(this.fsAPI);
        this.onPlayerReady.emit(this.api);
    }
    onChangeFullscreen(fsState) {
        if (!this.fsAPI.nativeFullscreen) {
            this.isFullscreen = fsState;
            this.zIndex = fsState ? VgUtils.getZIndex().toString() : 'auto';
        }
        else {
            this.isNativeFullscreen = fsState;
        }
    }
    onHideControls(hidden) {
        this.areControlsHidden = hidden;
    }
    ngOnDestroy() {
        this.subscriptions.forEach(s => s.unsubscribe());
    }
};
VgPlayer.ctorParameters = () => [
    { type: ElementRef },
    { type: VgAPI },
    { type: VgFullscreenAPI },
    { type: VgControlsHidden }
];
__decorate([
    HostBinding('class.fullscreen')
], VgPlayer.prototype, "isFullscreen", void 0);
__decorate([
    HostBinding('class.native-fullscreen')
], VgPlayer.prototype, "isNativeFullscreen", void 0);
__decorate([
    HostBinding('class.controls-hidden')
], VgPlayer.prototype, "areControlsHidden", void 0);
__decorate([
    HostBinding('style.z-index')
], VgPlayer.prototype, "zIndex", void 0);
__decorate([
    Output()
], VgPlayer.prototype, "onPlayerReady", void 0);
__decorate([
    Output()
], VgPlayer.prototype, "onMediaReady", void 0);
__decorate([
    ContentChildren(VgMedia)
], VgPlayer.prototype, "medias", void 0);
VgPlayer = __decorate([
    Component({
        selector: 'vg-player',
        encapsulation: ViewEncapsulation.None,
        template: `<ng-content></ng-content>`,
        providers: [VgAPI, VgFullscreenAPI, VgControlsHidden],
        styles: [`
        vg-player {
            font-family: 'videogular';
            position: relative;
            display: flex;
            width: 100%;
            height: 100%;
            overflow: hidden;
            background-color: black;
        }
        vg-player.fullscreen {
            position: fixed;
            left: 0;
            top: 0;
        }
        vg-player.native-fullscreen.controls-hidden {
            cursor: none;
        }
    `]
    })
    // tslint:disable:component-class-suffix
    // tslint:disable:no-output-on-prefix
], VgPlayer);
export { VgPlayer };
